File: apps/api/app/schemas/invoice.py
from decimal import Decimal

from pydantic import BaseModel

from .base import BaseSchema, MoneySerializerMixin, TimestampMixin
from .line_item import LineItemInInvoice


class InvoiceBase(BaseSchema):
    """Base Invoice fields"""

    id: int
    campaign_id: int


class InvoiceSummary(MoneySerializerMixin, InvoiceBase):
    """Invoice summary in Campaign detail"""

    total_actual: Decimal
    total_adjustments: Decimal
    total_billable: Decimal
    line_items_count: int


class InvoiceListItem(MoneySerializerMixin, InvoiceBase):
    """GET /api/v1/invoices list item"""

    campaign_name: str
    total_billable: Decimal
    line_items_count: int


class InvoiceDetail(MoneySerializerMixin, InvoiceBase, TimestampMixin):
    """GET /api/v1/invoices/{id} detail"""

    campaign_name: str
    line_items: list[LineItemInInvoice]
    total_actual: Decimal
    total_adjustments: Decimal
    total_billable: Decimal


class InvoiceListResponse(BaseModel):
    """GET /api/v1/invoices response"""

    invoices: list[InvoiceListItem]
    total: int
    limit: int
    offset: int

File: apps/web/src/types/invoice.ts
export interface LineItemInInvoice {
  id: number;
  campaign_id: number;
  name: string;
  booked_amount: string;
  actual_amount: string;
  adjustments: string;
  billable_amount: string;
  invoice_line_item_id: number;
}

export interface InvoiceListItem {
  id: number;
  campaign_id: number;
  campaign_name: string;
  total_billable: string;
  line_items_count: number;
}

export interface InvoiceDetail {
  id: number;
  campaign_id: number;
  campaign_name: string;
  created_at: string;
  updated_at: string;
  line_items: LineItemInInvoice[];
  total_actual: string;
  total_adjustments: string;
  total_billable: string;
}

export interface InvoiceListResponse {
  invoices: InvoiceListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface InvoiceLineItemResponse {
  id: number;
  invoice_id: number;
  line_item_id: number;
  actual_amount: string;
  adjustments: string;
  billable_amount: string;
  updated_at: string;
}

export interface AdjustmentItem {
  invoice_line_item_id: number;
  adjustments: string;
}

export interface BatchAdjustmentsUpdate {
  updates: AdjustmentItem[];
}

export interface BatchAdjustmentsResponse {
  invoice_id: number;
  updated: InvoiceLineItemResponse[];
}
